[MM-17390] Rate limited posts fail to load threads

cpoile opened the pull request:
#### Summary
- Rate-limited posts are batched, and those posts skip the `completePostReceived` path, which includes getting the post's root. If the user then clicks on the thread, they would receive the message:
![image](https://user-images.githubusercontent.com/1490756/63042256-9a534b80-be97-11e9-9823-d329ac730f2e.png)
- This fix uses the same strategy as the `getProfilesAndStatusesForPosts` function, which is also run for the list of rate-limited posts.

#### Ticket Link
- [MM-17390](https://mattermost.atlassian.net/browse/MM-17390)

#### Checklist
[Place an '[x]' (no spaces) in all applicable fields. Please remove unrelated fields.]
- [x]  Ran `make check-style` to check for style errors (required for all pull requests)
- [x] Ran `make test` to ensure unit tests passed
- [x] Ran `make flow` to ensure type checking passed
- [x] Added or updated unit tests (required for all new features)
- [ ] All new/modified APIs include changes to the [JavaScript driver](https://github.com/mattermost/mattermost-redux/blob/master/src/client/client4.js)

#### Test Information
This PR was tested on: Mac, Chrome

lieut-data commented on src/actions/posts.js:
```diff
@@ -759,6 +760,28 @@ export function getPostsAround(channelId, postId, perPage = Posts.POST_CHUNK_SIZ
+
+    const promises = [];
+
+    posts.forEach((post) => {
```
Aha, here's the complexity. Is it true that for each post we would have "normally" received, we call `getPostThread`, and so doing them all now isn't any "worse" as such? Slightly worried about slamming the server in this scenario, and wondering if we need a `getPostsThreads` instead.

cpoile replied:
Yep, for every normal post we would do this (and more). Here, we only get the thread if we don't have it in the store already. I think a `getPostsThreads` wouldn't be any different (we've got to get each of those threads somehow, right?).

When Christopher was making the rate-limiting changes, he said the performance hit from `getProfilesAndStatusesForPosts` isn't significant, and it's got a lot more going on than `getThreadsForPosts`.  (https://github.com/mattermost/mattermost-webapp/pull/2041)

lieut-data commented on src/actions/posts.js:
```diff
@@ -759,6 +760,28 @@ export function getPostsAround(channelId, postId, perPage = Posts.POST_CHUNK_SIZ
+            return;
+        }
+        const rootPost = Selectors.getPost(getState(), post.root_id);
+        if (post.root_id && !rootPost) {
```
nit: we've already checked `post.root_id` above:
```suggestion
        if (!rootPost) {
```

cpoile replied:
d'oh! Thanks. :)

enahum commented on src/actions/posts.js:
```diff
@@ -759,6 +760,28 @@ export function getPostsAround(channelId, postId, perPage = Posts.POST_CHUNK_SIZ
+        }
+        const rootPost = Selectors.getPost(getState(), post.root_id);
+        if (!rootPost) {
+            promises.push(PostActions.getPostThread(post.root_id)(dispatch, getState));
```
Nit:
```suggestion
            promises.push(dispatch(PostActions.getPostThread(post.root_id));
```

cpoile replied:
@enahum Cool, thanks!

cpoile commented on src/actions/posts.js:
```diff
@@ -759,6 +760,28 @@ export function getPostsAround(channelId, postId, perPage = Posts.POST_CHUNK_SIZ
 
+// getThreadsForPosts is intended for an array of posts that have been batched
+// (see the actions/websocket_actions/handleNewPostEvents function in the webapp)
+export function getThreadsForPosts(posts, dispatch, getState) {
```
@enahum I was going off of the function just below it (`getProfilesAndStatusesForPosts`) since they're both called in the same context (see: https://github.com/mattermost/mattermost-webapp/pull/3449). Does that change your suggestion?

enahum replied:
I would change it from `getThreadsForPosts(posts, myDispatch, myGetState);` to `myDispatch(getThreadsForPosts(posts));

cpoile replied:
@enahum Would you replace `return Promise.all(promises)` with...?

cpoile replied:
Changes made, thanks for the improvements @enahum !

enahum replied:
is ok to leave the promise.all for now as we don't really use the result of it anywhere, but for the future if you need the results you can go with an `allSettle` kind of Promise resolution

enahum commented on src/actions/posts.test.js:
```diff
@@ -1556,4 +1556,40 @@ describe('Actions.Posts', () => {
+            });
+
+            it('null', async () => {
+                await Actions.getThreadsForPosts(null, dispatch, getState);
```
```suggestion
                await dispatch(Actions.getThreadsForPosts(null));
```

enahum commented on src/actions/posts.test.js:
```diff
@@ -1556,4 +1556,40 @@ describe('Actions.Posts', () => {
+            it('array of posts', async () => {
+                const posts = [post1, post2];
+
+                await Actions.getThreadsForPosts(posts, dispatch, getState);
```
```suggestion
                await dispatch(Actions.getThreadsForPosts(posts));
```

cpoile commented:
I updated the tests and then realized they weren't really testing anything. So I rewrote them. I would appreciate your eyes on the new code, thanks!

@enahum I have one other question... The test for the webapp portion is failing: https://github.com/mattermost/mattermost-webapp/pull/3449 . I think it's because we changed `getThreadsForPosts` to be a HOF, and now the tests in the webapp fail with `Error: Actions must be plain objects. Use custom middleware for async actions.` at line 224 of `websocket_actions.test.jsx`.

I've tried adding redux-thunk to the configureStore: `const testStore = configureStore([thunk]);` but that doesn't work. I'm sorry to ask, but I've spent an hour on it and I just don't understand redux well enough yet to figure it out. How would you fix this?

enahum commented:
@cpoile this looks good, the issue on the webapp is probably cause you need to update the mm-redux reference in the package(-lock).json files

cpoile commented:
@enahum I mean the tests are failing on my local build, which is updated (using npm run dev:watch in the redux repo).

enahum commented:
@cpoile let me try it locally and see

saturninoabril commented on src/actions/posts.js:
```diff
@@ -11,6 +11,7 @@ import {getCurrentChannelId, getMyChannelMember as getMyChannelMemberSelector} f
 import {getCustomEmojisByName as selectCustomEmojisByName} from 'selectors/entities/emojis';
 import {getConfig} from 'selectors/entities/general';
 import * as Selectors from 'selectors/entities/posts';
+import * as PostActions from 'actions/posts';
```
Is this necessary?  It's importing the same file, right?

cpoile replied:
good point, thanks!

saturninoabril commented on src/actions/posts.js:
```diff
@@ -759,6 +760,32 @@ export function getPostsAround(channelId, postId, perPage = Posts.POST_CHUNK_SIZ
+
+            const rootPost = Selectors.getPost(state, post.root_id);
+            if (!rootPost) {
+                promises.push(dispatch(PostActions.getPostThread(post.root_id)));
```
Can this be
```suggestion
                promises.push(dispatch(getPostThread(post.root_id)));
```

cpoile commented:
Removing cherry-pick because this is for 5.15, not 5.14.

amyblais commented:
@cpoile Please cherry-pick to v5.15.

lieut-data commented:
/cherry-pick v5.15

mattermod commented:
Error trying doing the automated Cherry picking. Please do this manually

```
+++ Updating remotes...
Fetching upstream
Fetching origin
+++ Updating remotes done...
!!! 'upstream/v5.15' not found. The second argument should be something like upstream/release-0.21.
    (In particular, it needs to be a valid, existing remote branch that I can 'git checkout'.)

```

lieut-data commented:
/cherry-pick 5.15

mattermod commented:
Error trying doing the automated Cherry picking. Please do this manually

```
+++ Updating remotes...
Fetching upstream
Fetching origin
+++ Updating remotes done...
!!! 'upstream/5.15' not found. The second argument should be something like upstream/release-0.21.
    (In particular, it needs to be a valid, existing remote branch that I can 'git checkout'.)

```

lieut-data commented:
(Tried to do this automatically, then realized it needs special handling re: redux.)

cpoile commented:
/cherry-pick release-5.15

mattermod commented:
Error trying doing the automated Cherry picking. Please do this manually

```
+++ Updating remotes...
Fetching upstream
Fetching origin
+++ Updating remotes done...
+++ Creating local branch automated-cherry-pick-of-#898-upstream-release-5.15-1566320816
Branch 'automated-cherry-pick-of-#898-upstream-release-5.15-1566320816' set up to track remote branch 'release-5.15' from 'upstream'.
+++ Downloading patch to /tmp/898.patch (in case you need to do this again)

+++ About to attempt cherry pick of PR. To reattempt:
  $ git am -3 /tmp/898.patch

Applying: * fixes batched posts not loading threads
Applying: Update src/actions/posts.js
Applying: Update src/actions/posts.js
Applying: * PR comments
Using index info to reconstruct a base tree...
M	src/actions/posts.js
Falling back to patching base and 3-way merge...
Auto-merging src/actions/posts.js
CONFLICT (content): Merge conflict in src/actions/posts.js
Patch failed at 0004 * PR comments
Use 'git am --show-current-patch' to see the failed patch
When you have resolved this problem, run "git am --continue".
If you prefer to skip this patch, run "git am --skip" instead.
To restore the original branch and stop patching, run "git am --abort".

+++ Conflicts detected:

UU src/actions/posts.js

+++ Aborting in-progress git am.

+++ Returning you to the master branch and cleaning up.

```
